### PR TITLE
[Merged by Bors] - fix: make the long file linter reject "too large" values

### PR DESCRIPTION
**Pull request by adomani**
Reported [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Splitting.20Mathlib.2FData.2FFinset.2FBasic/near/473746566).

The PR corrects the checks to only allow "small" increases to the value of the option.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/Hom/Lattice.lean [hunk 1]
```diff
@@ -1777,4 +1777,4 @@ def withTopWithBot' [BoundedOrder β] (f : LatticeHom α β) :
 
 end LatticeHom
 
-set_option linter.style.longFile 1800
+set_option linter.style.longFile 1900
```
Comment (adomani) [inline Mathlib/Order/Hom/Lattice.lean:1780]:
```suggestion
set_option linter.style.longFile 1800
```

**Comment by grunweg** [comment]
Thank you for working on this - and I am really happy there are thorough tests for this now. This logic is complex enough that holding it all in my head is error-prone, as we saw :-)

I just saw another pre-existing issue: there are two more references to 1500 in the file. Feel free to file a PR with just that change, label it `easy` and `maintainer merge` on my behalf. (I'll be without github access until the end of the week.)

File: Mathlib/Tactic/Linter/Lint.lean [hunk 2]
```diff
@@ -354,6 +354,7 @@ def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
   if let some init := stx.getTailPos? then
     -- the last line: we subtract 1, since the last line is expected to be empty
     let lastLine := ((← getFileMap).toPosition init).line
+    -- in this case, the file has an allowed length, and the linter option is unnecessarily set
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:357]: Good idea to add a comment!
```suggestion
    -- In this case, the file has an allowed length, and the linter option is unnecessarily set.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 3]
```diff
@@ -365,12 +366,23 @@ def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
     -- `candidate` is necessarily bigger than `lastLine` and hence bigger than `defValue`
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:366]:
```
    -- Note that either `lastLine ≤ defValue` and `defValue = linterBound` hold or
    -- `candidate` is necessarily bigger than `lastLine` and hence bigger than `defValue`.

Comment (grunweg) [reply Mathlib/Tactic/Linter/Lint.lean:366]: Note to self: triple-check this logic; I am not fully convinced any more that these are *all* possible cases.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 4]
```diff
@@ -365,12 +366,23 @@ def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
...
     let candidate := (lastLine / 100) * 100 + 200
     let candidate := max candidate defValue
-    if linterBound < lastLine then
+    -- in this case, the file is longer than the default and also than what the option says
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:369]:
```suggestion
    -- In this case, the file is longer than the default and also than what the option says.

File: test/LongFile.lean [hunk 5]
```diff
@@ -20,49 +20,61 @@ The current value of 1500 does not exceed the allowed bound.
 Please, remove the `set_option linter.style.longFile 1500`.
 -/
 #guard_msgs in
--- Do not allow setting a "small" `longFile` linter option
+-- Do not allow setting a `longFile` linter option if the file does not exceed the `defValue`
 set_option linter.style.longFile 1500
```
Comment (grunweg) [inline test/LongFile.lean:24]: What do you think about setting this to the default value instead? I wouldn't want to change that default and break the test... (similarly, auditing this file for occurrences of the number 1500 seems useful)

Comment (grunweg) [reply test/LongFile.lean:24]: Can you add another test setting the option to something *smaller* than the default (i.e., 100 lines or so) --- the goal being to test the very first error message. That seems to be missing.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 6]
```diff
@@ -354,6 +354,7 @@ def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
   if let some init := stx.getTailPos? then
     -- the last line: we subtract 1, since the last line is expected to be empty
     let lastLine := ((← getFileMap).toPosition init).line
+    -- in this case, the file has an allowed length, and the linter option is unnecessarily set
     if lastLine ≤ defValue && defValue < linterBound then
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:358]: Can you check that some test fails if you remove the second condition in the `if`? If not, please add a test :-)

File: Mathlib/Tactic/Linter/Lint.lean [hunk 7]
```diff
@@ -365,12 +366,23 @@ def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
     -- `candidate` is necessarily bigger than `lastLine` and hence bigger than `defValue`
     let candidate := (lastLine / 100) * 100 + 200
     let candidate := max candidate defValue
-    if linterBound < lastLine then
+    -- in this case, the file is longer than the default and also than what the option says
+    if defValue ≤ linterBound && linterBound < lastLine then
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:370]: It seems we have a test for this, when the default value is exceeded. Can you also add one where it's equal? (I don't want to have to worry about off-by-one errors, but have a test having my back.)

File: Mathlib/Tactic/Linter/Lint.lean [hunk 8]
```diff
@@ -365,12 +366,23 @@ def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
...
       logWarningAt stx <| .tagged linter.style.longFile.name
         m!"This file is {lastLine} lines long, but the limit is {linterBound}.\n\n\
           You can extend the allowed length of the file using \
           `set_option linter.style.longFile {candidate}`.\n\
           You can completely disable this linter by setting the length limit to `0`."
+    else
+    -- finally, the file exceeds the default value, but not the option: we only allow the value
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:377]: You *could* phrase this check differently:
- check early if the default value is not a multiple of 100 (which seems desirable regardless, right)
- then, just compare candidate and linterBound + 200 here
Given the first change, you could then make the error message clearer: the limit is *too high* and should be lowered.

File: test/LongFile.lean [hunk 9]
```diff
@@ -20,49 +20,61 @@ The current value of 1500 does not exceed the allowed bound.
...
+`set_option linter.style.longFile 200`.
+You can completely disable this linter by setting the length limit to `0`.
 -/
 #guard_msgs in
--- First, we silence the linter, so that we can set a default value smaller than 1500.
-set_option linter.style.longFile 0 in
--- If we set the allowed bound for the `longFile` linter that is too large,
--- the linter tells us to use a smaller bound.
-set_option linter.style.longFile 1700 in
+-- Check that a value different from `candidate` or `candidate - 100` value is not allowed
```
Comment (grunweg) [inline test/LongFile.lean:75]: As I said above: a test which makes the allowed option far too large would also be nice. Thanks!

**Review by grunweg — COMMENTED** [review]
I took a look again. I remember how sufficiently subtle this is - hence, my review is focusing on test coverage. If there are tests for every edge case and we agree the linter behaves sensible there, I'll be more confident the logic is actually correct :-)

I also have a few minor tweaks, as usual. Thanks for continuing to work on this. I appreciate that landing this one is a bit of a nightmare, because
- you have to page all the fiddly details into your head, and
- the reviewer also needs to do so,
- ideally quickly enough for you to not lose that context.
Perhaps adding more tests can break this loop...

**Comment by grunweg** [comment]
awaiting-author

File: test/LongFile.lean (hunk 5 above)
Comment (adomani) [reply test/LongFile.lean:24]: > Can you add another test setting the option to something _smaller_ than the default (i.e., 100 lines or so) --- the goal being to test the very first error message. That seems to be missing.

The very first test of the file verifies that 1500 is not allowed.  I added a test for 1999 at the end of the file, after setting the default value to 2000, to make sure that the check uses `defValue`, rather than `1500`.  I did not add it next to the one that is already checking this, since it would involve re-checking manually that the claimed error that is not tested is still up to date.

Comment (adomani) [reply test/LongFile.lean:24]: By the way, there are three `1500`s in the linter file: two are in doc-strings and one is the default value of the linter.  The rest of the file uses the variables, rather than the number.

File: test/LongFile.lean (hunk 9 above)
Comment (adomani) [reply test/LongFile.lean:75]: I added a test at the end, setting the `defValue` to 400 and the `linterBound` to 5000.

**Review by adomani — COMMENTED** [review]
Michael, thank you for the review!  I added some tests.

File: Mathlib/Tactic/Linter/Lint.lean (hunk 6 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Lint.lean:358]: Thanks!

File: Mathlib/Tactic/Linter/Lint.lean (hunk 7 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Lint.lean:370]: Fair enough: I'm also won't object to not verifying that situation super closely.

File: test/LongFile.lean (hunk 5 above)
Comment (grunweg) [reply test/LongFile.lean:25]: I found one case clearer with replacing it - but agree on the other ones!

**Review by grunweg — APPROVED** [review]
I just re-read everything again and pushed a commit with a few last tweaks. (In particular, I removed the "you can disable this linter" message from the newly added branch: the user decided to configure this lint, so there seems to be no need to tell them about removing it, right? That lint only applies to mathlib anyway, which would not want to disallow the linter. Feel free to revert that one, if you disagree.)
maintainer merge

**Comment by jcommelin** [comment]
CI is unhappy atm

**Comment by grunweg** [comment]
Fixed the CI failure.

**Comment by kim-em** [comment]
bors merge
